flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/NestedPositionUtil.java

```diff
@@ -0,0 +1,214 @@
+     * @return {@link RowPosition} contains collections row count and isNull array.
+     */
+    public static RowPosition calculateRowOffsets(
+            Field field, int[] fieldDefinitionLevels, int[] fieldRepetitionLevels) {
+        int rowDefinitionLevel = field.getDefinitionLevel();
+        int rowRepetitionLevel = field.getRepetitionLevel();
+        int nullValuesCount = 0;
+        BooleanArrayList nullRowFlags = new BooleanArrayList();
+        for (int i = 0; i < fieldDefinitionLevels.length; i++) {
+            if (fieldRepetitionLevels[i] > rowRepetitionLevel) {
+                throw new IllegalStateException(
+                        format(
+                                "In parquet's row type field repetition level should not larger than row's repetition level. "
+                                        + "Row repetition level is %s, row field repetition level is %s.",
+                                rowRepetitionLevel, fieldRepetitionLevels[i]));
+            }
+
+            if (fieldDefinitionLevels[i] >= rowDefinitionLevel) {
+                // current row is defined and not empty
+                nullRowFlags.add(false);
+            } else {
+                // current row is null
+                nullRowFlags.add(true);
+                nullValuesCount++;
+            }
+        }
+        if (nullValuesCount == 0) {
+            return new RowPosition(Optional.empty(), fieldDefinitionLevels.length);
+        }
+        return new RowPosition(Optional.of(nullRowFlags.toBooleanArray()), nullRowFlags.size());
+    }
+
+    /**
+     * Calculate the collection's offsets according to column's max repetition level, definition
+     * level, value's repetition level and definition level. Each collection (Array or Map) has four
+     * situation:
+     * <li>Collection is not defined, because optional parent fields is null, this is decided by its
+     *     parent's repetition level
+     * <li>Collection is null
+     * <li>Collection is defined but empty
+     * <li>Collection is defined and not empty. In this case offset value is increased by the number
+     *     of elements in that collection
+     *
+     * @param field field that contains array/map column message include max repetition level and
+     *     definition level.
+     * @param definitionLevels int array with each value's repetition level.
+     * @param repetitionLevels int array with each value's definition level.
+     * @return {@link CollectionPosition} contains collections offset array, length array and isNull
+     *     array.
+     */
+    public static CollectionPosition calculateCollectionOffsets(
+            Field field, int[] definitionLevels, int[] repetitionLevels) {
+        int collectionDefinitionLevel = field.getDefinitionLevel();
+        int collectionRepetitionLevel = field.getRepetitionLevel() + 1;
+        int offset = 0;
+        int valueCount = 0;
+        LongArrayList offsets = new LongArrayList();
```

Comment:
Take a look to `org.apache.flink.runtime.util.LongArrayList`.

Reply:
done